Forward "express:usage-record" to strong-pm

bajtos opened the pull request:
If we have an ipc channel to our parent, we forward express usage records (`express:usage-record`) from strong-agent to the parent.

/to @sam-github please review

Close strongloop-internal/scrum-loopback#125

bajtos commented:
@sam-github could you please advise what am I doing wrong? It seems there are multiple instances of `strong-agent`, the instance used by the probe to emit events is different from the instance returned by `agent()`.

Test file:

```js
// metrics.js
var http = require('http');

var app = require('express')();
app.use(require('strong-express-metrics')());

app.listen(3000, function() {
  http.get(
    {
      host: this.address().address,
      port: this.address().port,
      path: '/',
    },
    function(res) {
      res.resume();
    });
});
```

You need to `npm install strongloop/strong-express-metrics#2d32c9c` in the directory where `metrics.js` is placed and patched strong-agent in the supervisor. Then run the following command:

```
DEBUG=* ~/src/strong/supervisor/bin/sl-run.js --cluster 1 metrics.js
```

sam-github commented on lib/config.js:
```diff
@@ -319,6 +320,8 @@ config.start = function start() {
     control.start({ size: config.size });
   });
 
+  setupExpresUsageForwarding(runctl.parentCtl);
```
"Express"

sam-github commented on lib/config.js:
```diff
@@ -340,6 +343,17 @@ config.start = function start() {
   }
+
+  function setupExpresUsageForwarding(parentCtl) {
+    var debugExpress = require('./debug')('express-usage');
```
We tend to use 1 debug per file, scoped on the file, and there is one already in config. If you use it, you can remove this function, and don't have to spread the code across two places.

sam-github commented on lib/config.js:
```diff
@@ -340,6 +343,17 @@ config.start = function start() {
+
+  function setupExpresUsageForwarding(parentCtl) {
+    var debugExpress = require('./debug')('express-usage');
+    agent().on(
```
if there is not parentctl channel, you don't have to do the agent().on(), I don't think, because there is no destination.

bajtos replied:
The idea is that since there won't be any unit-tests, I want to use the debug log to check that the records are emitted/forwarded. Even in the case when there is no parentctl channel.

sam-github replied:
I think the comment above must predate the conversation we just had in slack.

Anyhow, there should be unit tests, and once you pull this code into lib/express-stats.js, the debug will scoped on 'express-stats'. Not that you will need that for unit tests.

bajtos commented:
@sam-github I have reworked the PR per our yesterday's discussion. PTAL.

Jenkins runs only mocha tests, thus my new test is not run as part of the CI :(

bajtos commented:
The following two TAP tests are failing on my machine in both "master" and "feature/forward-express-metrics" branches. I assume the failure is not related to my changes.

**test/test-run-metrics.js**

```
not ok test/test-run-metrics.js ......................... 0/1
    Command: "/Users/bajtos/.nvm/v0.10.36/bin/node test-run-metrics.js"
    TAP version 13
    not ok 1 test/test-run-metrics.js
      ---
        exit:     143
        timedOut: true
        command:  "/Users/bajtos/.nvm/v0.10.36/bin/node test-run-metrics.js"
      ...

    1..1
    # tests 1
    # fail  1
```

**test/test-ipcctl-notifications.js**

```
assert.js:93
  throw new assert.AssertionError({
        ^
AssertionError: false == true
    at /Users/bajtos/src/strong/supervisor/test/test-ipcctl-notifications.js:176:5
    at Protocol.process (/Users/bajtos/src/strong/supervisor/node_modules/strong-control-channel/lib/message.js:111:5)
    at ChildProcess.self._onmessage (/Users/bajtos/src/strong/supervisor/node_modules/strong-control-channel/lib/process.js:21:20)
    at ChildProcess.emit (events.js:98:17)
    at handleMessage (child_process.js:322:10)
    at Pipe.channel.onread (child_process.js:349:11)
not ok test/test-ipcctl-notifications.js ................ 0/1
    Command: "/Users/bajtos/.nvm/v0.10.36/bin/node test-ipcctl-notifications.js"
    TAP version 13
    not ok 1 test/test-ipcctl-notifications.js
      ---
        exit:    8
        stderr:  |
          assert.js:93
            throw new assert.AssertionError({
                  ^
          AssertionError: false == true
              at /Users/bajtos/src/strong/supervisor/test/test-ipcctl-notifications.js:176:5
              at Protocol.process (/Users/bajtos/src/strong/supervisor/node_modules/strong-control-channel/lib/message.js:111:5)
              at ChildProcess.self._onmessage (/Users/bajtos/src/strong/supervisor/node_modules/strong-control-channel/lib/process.js:21:20)
              at ChildProcess.emit (events.js:98:17)
              at handleMessage (child_process.js:322:10)
              at Pipe.channel.onread (child_process.js:349:11)
        command: "/Users/bajtos/.nvm/v0.10.36/bin/node test-ipcctl-notifications.js"
      ...

    1..1
    # tests 1
    # fail  1
```

rmg commented on test/test-run-express-metrics.js:
```diff
@@ -0,0 +1,34 @@
+var helper = require('./helper');
+var run = helper.runWithControlChannel;
+
+helper.pass = true; // Use tap, not this check.
```
You'll need to add something to skip these tests when run under mocha. Currently CI is failing because it tries to run this code but `helper.runWithControlChannel` isn't defined because helper.js returns before exporting it.

Something like this near the top of the file:
```js
if (helper.skip()) return;
```

bajtos replied:
Good catch, I added a new commit 5410fbb that should fix the problem.

sam-github commented on bin/sl-run.js:
```diff
@@ -31,6 +31,8 @@ config.sendMetrics();
 if(!config.clustered) {
   console.log('supervisor running without clustering (unsupervised)');
+} else {
+  config.sendExpressMetrics();
```
You can call this unconditionally, like sendMetrics() and just after it, and *inside* sendExpressMetrics make it a no-op if `! this.clustered`.

bajtos replied:
`sendExpressMetrics` does not have a reference to the config, thus I can't put the check there. I'll do the check in `config.start`.

sam-github commented on lib/config.js:
```diff
@@ -11,6 +11,7 @@ var fs = require('fs');
 var open = require('fs').openSync;
 var path = require('path');
 var sendMetrics = require('./metrics');
+var sendExpressMetrics = require('./express-metrics');
```
I think `express-record` and `sendExpressRecord()` would make more sense for this. We use metrics in various places in docs and code to refer specifically to data that can be expressed as key/value pairs sendable using statsd, and other forms, and to say things like "a metrics license is required". The word "metrics" isn't used for other purposes, and it would be better not to start.

bajtos replied:
I'll use `express-records` and `sendExpressRecords()`, as the code is sending more than one record.

sam-github commented on lib/express-metrics.js:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  cluster.on('fork', function(worker) {
+    worker.on('message', function(msg) {
```
This function could be called `forwardRecordToParent`, if you wanted, for parallelism with the other path.

sam-github commented on test/express-app/package.json:
```diff
@@ -1,5 +1,6 @@
   "description": "express-app",
-  "version": "0.0.0"
+  "version": "0.0.0",
+  "main": "server.js"
```
Did you find this necessary? server.js is one of the defaults, it shouldn't have to be here. I can slc run this app without a main key.

bajtos replied:
I am using `require.resolve` to get an absolute path to the express app, and that does not work with `server.js`

sam-github commented on test/test-run-express-metrics.js:
```diff
@@ -0,0 +1,35 @@
+var helper = require('./helper');
```
This test times out for me, I guess it passes for you? I wonder if there is a race condition between starting sl-run, and starting to listen for messages from it (I don't have time to check now, just a thought, if you are on a mac linux scheduling could be a difference).

bajtos replied:
The code starting sl-run and listening for messages is based on what was in `test/test-run-metrics.js`. IIUC, the messages are send via `fd 3` of type `ipc`, so there shouldn't be any race condition?

bajtos replied:
Strangely enough, the test is not working on my machine today, I'll investigate.

sam-github commented:
I just noticed that there were some pre-existing failures. Sorry, I'll look at when I have a chance.

bajtos commented:
I have addressed all comments above. The rename screwed up the history, so I squashed the commits.

The test failure is a mystery to me. The new test used to fail on every run. Then I did `git rebase -i && git push -f` and suddenly they are passing again.

For the record, here is the full debug output of the situation when the test failed:

```
  strong-supervisor:test spawn: ["/Users/bajtos/src/strong/supervisor/bin/sl-run.js","--no-timestamp-workers","--no-timestamp-supervisor","--cluster=1","/Users/bajtos/src/strong/supervisor/test/express-app/server.js"] +0ms
  strong-control-channel:process:72345 parent 72345 attaching to child 72348 +7ms
pid:72348 worker:0 INFO strong-agent dashboard is at https://strongops.strongloop.com
pid:72348 worker:0 INFO strong-agent using collector https://127.0.0.1:8443
pid:72348 worker:0 INFO strong-agent v1.3.2 profiling app 'express-app' pid '72348'
pid:72348 worker:0 INFO strong-agent[72348] started profiling agent
pid:72348 worker:0 INFO supervisor starting (pid 72348)
pid:72348 worker:0 INFO strong-agent strong-agent using strong-cluster-control v2.0.0
pid:72348 worker:0 INFO supervisor reporting metrics to `internal:`
  strong-control-channel:process:72345 protocol notification #0 {"cmd":"started","pid":72348,"appName":"express-app","agentV... +262ms
pid:72348 worker:0 INFO strong-agent agent metrics license not found, local reporting disabled.
pid:72348 worker:0 Please contact sales@strongloop.com for assistance.
pid:72348 worker:0 INFO supervisor size set to 1
pid:72348 worker:0 INFO supervisor listening on 'runctl'
  strong-control-channel:process:72345 protocol notification #1 {"cmd":"fork","id":1,"pid":72349} +12ms
  strong-control-channel:process:72345 protocol notification #2 {"master":{"pid":72348,"setSize":1},"workers":[{"id":"1","pi... +0ms
pid:72348 worker:0 ERROR supervisor exiting with code 1
```

@sam-github PTAL and let me know if there is anything to change, fix or improve.

sam-github commented:
The test times out for me on ubuntu every time, can you try on Linux? I'll troubleshoot this if its just my machine or you can't repro, but if it fails on my linux, will likely fail on CI.

bajtos commented:
> The test times out for me on ubuntu every time, can you try on Linux?

What is the output of `node test/test-run-express-records.js`?

sam-github commented:
And code looks great, thanks. 

```
sam@samtu:~/w/sn/strong-supervisor ((5410fbb...)) % node ./test/test-run-express-metrics.js 
pid:28806 worker:0 INFO strong-agent API key not found, StrongOps dashboard reporting disabled.
pid:28806 worker:0 Generate configuration with:
pid:28806 worker:0     npm install -g strongloop
pid:28806 worker:0     slc strongops
pid:28806 worker:0 See http://docs.strongloop.com/strong-agent for more information.
pid:28806 worker:0 INFO supervisor starting (pid 28806)
pid:28806 worker:0 INFO supervisor reporting metrics to `internal:`
pid:28806 worker:0 INFO strong-agent v1.3.2 profiling app 'express-app' pid '28806'
pid:28806 worker:0 INFO strong-agent[28806] started profiling agent
pid:28806 worker:0 INFO supervisor size set to 1
pid:28806 worker:0 INFO supervisor listening on 'runctl'
pid:28806 worker:0 INFO supervisor started worker 1 (pid 28808)
pid:28806 worker:0 INFO supervisor resized to 1
pid:28808 worker:1 INFO strong-agent[28808] started profiling agent
pid:28808 worker:1 express-app listening on http://localhost:52455
TAP version 13
# express-metrics are forwarded via parentCtl
not ok 1 Timeout!
  ---
    file:   timers.js
    line:   112
    column: 15
    stack:  
      - |
        getCaller (/home/sam/w/sn/strong-supervisor/node_modules/tap/lib/tap-assert.js:418:17)
      - |
        assert (/home/sam/w/sn/strong-supervisor/node_modules/tap/lib/tap-assert.js:21:16)
      - |
        Function.fail (/home/sam/w/sn/strong-supervisor/node_modules/tap/lib/tap-assert.js:95:10)
      - |
        Test._testAssert [as fail] (/home/sam/w/sn/strong-supervisor/node_modules/tap/lib/tap-test.js:87:16)
      - |
        Test.timeout (/home/sam/w/sn/strong-supervisor/node_modules/tap/lib/tap-test.js:31:8)
      - |
        Timer.listOnTimeout [as ontimeout] (timers.js:112:15)
  ...

1..1
# tests 1
# fail  1
EXIT: 0
PASS
pid:28806 worker:0 ERROR supervisor exiting with code 2                                        

```

sam-github commented:
With DEBUG output:

```
sam@samtu:~/w/sn/strong-supervisor ((5410fbb...)) % DEBUG=* node ./test/test-run-express-metrics.js
  strong-supervisor:test spawn: ["/home/sam/w/sn/strong-supervisor/bin/sl-run.js","--no-timestamp-workers","--no-timestamp-supervisor","--cluster=1","/home/sam/w/sn/strong-supervisor/test/express-app/server.js"] +0ms
  strong-control-channel:process:28876 parent 28876 attaching to child 28878 +5ms
pid:28878 worker:0 INFO strong-agent API key not found, StrongOps dashboard reporting disabled.
pid:28878 worker:0 Generate configuration with:
pid:28878 worker:0     npm install -g strongloop
pid:28878 worker:0     slc strongops
pid:28878 worker:0 See http://docs.strongloop.com/strong-agent for more information.
pid:28878 worker:0 INFO supervisor starting (pid 28878)
pid:28878 worker:0 INFO supervisor reporting metrics to `internal:`
  strong-control-channel:process:28876 protocol notification #0 {"cmd":"started","pid":28878,"appName":"express-app","agentV... +185ms
pid:28878 worker:0 INFO strong-agent v1.3.2 profiling app 'express-app' pid '28878'
pid:28878 worker:0 INFO strong-agent[28878] started profiling agent
pid:28878 worker:0 INFO supervisor size set to 1
pid:28878 worker:0 INFO supervisor listening on 'runctl'
  strong-control-channel:process:28876 protocol notification #1 {"cmd":"fork","id":1,"pid":28880} +9ms
  strong-control-channel:process:28876 protocol notification #2 {"master":{"pid":28878,"setSize":1},"workers":[{"id":"1","pi... +1ms
pid:28878 worker:0 ERROR supervisor exiting with code 1

^C%                                                                       
```

bajtos commented:
@sam-github what's the plan? Should we merge this PR and fix the timing issue in the unit-tests later, or do you want to fix the timing issue before we land it?

sam-github commented:
merge now, fix later.

bajtos commented:
Ehm, I was too quick and forgot to clean up the history before merge. Sorry for that :(

sam-github commented:
Oops! I've merged some `fixup!`s, too, it happens.

rmg commented:
@slnode test please
